### PR TITLE
Implement the ES2015 version of Object.getPrototypeOf and add a test file for it

### PR DESCRIPTION
**Pull request by pmarkee**
JerryScript-DCO-1.0-Signed-off-by: Peter Marki marpeter@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 1]
```diff
@@ -94,6 +97,67 @@ ecma_builtin_object_dispatch_construct (const ecma_value_t *arguments_list_p, /*
...
+      arg = ecma_op_create_number_object (arg);
+    }
+    if (ecma_is_value_string (arg))
+    {
+      arg = ecma_op_create_string_object (&arg, 1);
+    }
+    if (ecma_is_value_null (arg) || ecma_is_value_undefined (arg))
+    {
+      ret_value = ecma_raise_type_error (ECMA_ERR_MSG ("Cannot coerce null or undefined to object"));
+    }
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: I think there is a `ecma_op_to_object` in `ecma-conversion.h`, why don't we use it?

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 2]
```diff
@@ -94,6 +97,67 @@ ecma_builtin_object_dispatch_construct (const ecma_value_t *arguments_list_p, /*
...
+/**
+ * [[GetPrototypeOf]]
+ *
+ * See also:
+ *          ES2015 9.1.1
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value
+ */
+static ecma_value_t
+ecma_builtin_object_object_get_prototype_of (ecma_value_t this_arg,
```
Comment (galpeter) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: Can't we just merge the new code path into the other `ecma_builtin_object_object_get_prototype_of` method? Do we really need to copy the whole function?

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 3]
```diff
@@ -109,15 +108,27 @@ ecma_builtin_object_object_get_prototype_of (ecma_value_t this_arg, /**< 'this'
 {
   JERRY_UNUSED (this_arg);
   ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+  bool was_object = ecma_is_value_object (arg);
 
   /* 1. */
-  if (!ecma_is_value_object (arg))
+  if (!was_object)
   {
+    #ifndef CONFIG_DISABLE_ES2015_BUILTIN
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: `#` macros always start at line start.

`# ifdef` is correct, but I don't like it

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 4]
```diff
@@ -109,15 +108,27 @@ ecma_builtin_object_object_get_prototype_of (ecma_value_t this_arg, /**< 'this'
...
+    if (ecma_is_value_null (arg) || ecma_is_value_undefined (arg))
+    {
+      ret_value = ecma_op_to_object (arg);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: I don't understand this line.

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 5]
```diff
@@ -109,15 +108,25 @@ ecma_builtin_object_object_get_prototype_of (ecma_value_t this_arg, /**< 'this'
...
   {
+#ifndef CONFIG_DISABLE_ES2015_BUILTIN
+    ret_value = ecma_op_to_object (arg);
+    if (ECMA_IS_VALUE_ERROR (ret_value))
+    {
+      return ret_value;
+    }
+    arg = ret_value;
+#else /* CONFIG_DISABLE_ES2015_BUILTIN */
     ret_value = ecma_raise_type_error (ECMA_ERR_MSG ("Argument is not an object."));
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: `return ecma_raise_type_error...`

I think this makes the `if (ecma_is_value_object (arg))` below unnecessary.

File: jerry-core/ecma/builtin-objects/ecma-builtin-object.c [hunk 6]
```diff
@@ -109,15 +108,25 @@ ecma_builtin_object_object_get_prototype_of (ecma_value_t this_arg, /**< 'this'
...
   JERRY_UNUSED (this_arg);
   ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+  bool was_object = ecma_is_value_object (arg);
 
   /* 1. */
-  if (!ecma_is_value_object (arg))
+  if (!was_object)
   {
+#ifndef CONFIG_DISABLE_ES2015_BUILTIN
+    ret_value = ecma_op_to_object (arg);
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-object.c]: `arg = ecma_op_to_object (arg);` makes the `arg = ret_value;` below unnecessary.

**Pull request merged**